geom/src/polyline.rs

```diff
@@ -519,24 +519,32 @@ impl PolyLine {
         Ok(result)
     }
 
-    /// The resulting polygon is manually triangulated and may not have a valid outer Ring (but it
-    /// usually does).
+    /// This produces a `Polygon` with a valid `Ring`. It may crash if this polyline was made with
+    /// `unchecked_new`
     pub fn make_polygons(&self, width: Distance) -> Polygon {
```

Comment:
So what's going on here? Calling `to_outline` on rings does `PolyLine::unchecked_new`, which ignores the first = last point that'd normally be illegal. This is used for drawing a thick border around a ring:
![Screenshot from 2022-09-03 17-24-12](https://user-images.githubusercontent.com/1664407/188279578-4f530112-f609-4236-b92a-6d76887c033a.png)
The result right now isn't a `Polygon`. But I think it wouldn't be too hard for it to be. We're easily making a ring for the outer part of the polyline. The inner part isn't trimming the line segments back at that last intersection yet, causing the funky corner. But it wouldn't be hard to specialize this logic for `Ring`. `to_outline` should produce an outer and inner ring. It might fail if the original ring is too small, and the inner ring ends up collapsing, so the API would have to change.

Also related: https://github.com/lelongg/geo-offset could maybe help here. I haven't tried it in a while. https://github.com/georust/geo/issues/641 is also relevant.

Reply:
Just a thought... would things get really crazy if the polygon created by `to_outline` wasn't centered on the polygon, and instead its interior ring was the exterior of the outlined polygon.

That is, the generated outline polygon would be strictly *outside* of its input.

I could imagine this being hairy - like maybe there'd be little pixel gaps between the input and its outline. Anyway, just spit-ballin.

Reply:
Ah, I see! Instead of having to both shrink and expand a `Ring` to make an outline, we could just expand and use the original ring as the inner hole. The shrinking case probably has more edge cases with some line segments disappearing entirely.

From a design perspective, I'd actually prefer if the outlines for polygons were entirely contained in the original polygon. The half-straddling approach now is particularly weird for hovering over lane polygons:
![Uploading Screenshot from 2022-09-12 11-45-09.png…]()